OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,779 @@
+      }
+
+      for (Marker& marker : command_buffer_state.markers) {
+        if (marker.slot_index.has_value()) {
+          slots_to_reset.push_back(marker.slot_index.value());
+          marker.slot_index.reset();
+        }
+      }
+    }
+    if (!slots_to_reset.empty()) {
+      timer_query_pool_->ResetQuerySlots(device, slots_to_reset);
+    }
+  }
+
+ private:
+  enum class MarkerType { kDebugMarkerBegin = 0, kDebugMarkerEnd };
+
+  struct Marker {
+    MarkerType type;
+    std::optional<uint32_t> slot_index;
+    std::optional<std::string> text;
+    std::optional<internal::Color> color;
+    bool cut_off;
+  };
+
+  struct QueueMarkerState {
+    std::stack<internal::MarkerState> marker_stack;
+  };
+
+  struct CommandBufferState {
+    std::optional<uint32_t> command_buffer_begin_slot_index;
+    std::optional<uint32_t> command_buffer_end_slot_index;
+    std::vector<Marker> markers;
+    uint32_t local_marker_stack_size;
+  };
+
+  uint32_t RecordTimestamp(VkCommandBuffer command_buffer,
+                           VkPipelineStageFlagBits pipeline_stage_flags) {
+    VkDevice device;
+    {
+      absl::ReaderMutexLock lock(&mutex_);
+      CHECK(command_buffer_to_device_.contains(command_buffer));
+      device = command_buffer_to_device_.at(command_buffer);
+    }
+
+    VkQueryPool query_pool = timer_query_pool_->GetQueryPool(device);
+
+    uint32_t slot_index;
+    bool found_slot = timer_query_pool_->NextReadyQuerySlot(device, &slot_index);
+    CHECK(found_slot);
+    dispatch_table_->CmdWriteTimestamp(command_buffer)(command_buffer, pipeline_stage_flags,
+                                                       query_pool, slot_index);
+
+    return slot_index;
+  }
+
+  std::vector<internal::QueueSubmission> PullCompletedSubmissions(VkDevice device,
+                                                                  VkQueryPool query_pool) {
+    std::vector<internal::QueueSubmission> completed_submissions = {};
+
+    absl::WriterMutexLock lock(&mutex_);
+    for (auto& [unused_queue, queue_submissions] : queue_to_submissions_) {
+      auto submission_it = queue_submissions.begin();
+      while (submission_it != queue_submissions.end()) {
+        const internal::QueueSubmission& submission = *submission_it;
+        if (submission.submit_infos.empty()) {
+          submission_it = queue_submissions.erase(submission_it);
+          continue;
+        }
+
+        bool erase_submission = true;
+        // Let's find the last command buffer in this submission, so first find the last
+        // submit info that has at least one command buffer.
+        // We test if for this command buffer, we already have a query result for its last slot
+        // and if so (or if the submission does not contain any command buffer) erase this
+        // submission.
+        auto submit_info_reverse_it = submission.submit_infos.rbegin();
+        while (submit_info_reverse_it != submission.submit_infos.rend()) {
```

Comment:
Do we always expect to find the submit info?

Reply:
Yes, otherwise, the `QueueSubmission` wouldn't expected to be written. I added a comment.